c++/src/kj/buffering.c++

```diff
@@ -0,0 +1,324 @@
+#include "buffering.h"
+#include <kj/debug.h>
+
+namespace kj {
+
+namespace _ {
+  static inline uint lg(uint value) {
+    // Compute floor(log2(value)).
+    //
+    // Undefined for value = 0.
+  #if _MSC_VER && !defined(__clang__)
+    unsigned long i;
+    auto found = _BitScanReverse(&i, value);
+    KJ_DASSERT(found);  // !found means value = 0
+    return i;
+  #else
+    return sizeof(uint) * 8 - 1 - __builtin_clz(value);
+  #endif
+  }
+}
+
+RetainableFdBufferRange::RetainableFdBufferRange(Badge<SplittableFdBufferRange>,
+    FdRangeInputBuffer& parent, ArrayPtr<byte> bytes, Array<AutoCloseFd> fds)
+    : parentOrRetainedBytes(&parent),
+      bytes(bytes),
+      fds(mv(fds)) {}
+
+RetainableFdBufferRange::RetainableFdBufferRange(Badge<SplittableFdBufferRange>,
+    Own<void> ownBytes, ArrayPtr<byte> bytes, Array<AutoCloseFd> fds)
+    : parentOrRetainedBytes(mv(ownBytes)),
+      bytes(bytes),
+      fds(mv(fds)) {}
+
+ArrayPtr<byte> RetainableFdBufferRange::getBytes() { return bytes; }
+ArrayPtr<AutoCloseFd> RetainableFdBufferRange::getFds() { return fds; }
+
+void RetainableFdBufferRange::retain() {
+  KJ_SWITCH_ONEOF(parentOrRetainedBytes) {
+    KJ_CASE_ONEOF(parent, FdRangeInputBuffer*) {
+      Array<byte> ownBytes;
+      if (bytes.size() > parent->bytes.size() * 0.8) {
```

Comment:
Hmmm, I'd take this `0.8` ratio as an option. In practice, this will probably be a useful knob. That said, can you make the parent responsible for this decision instead of this class?